File: doodlestudio/director/validate.py
"""Check a storyboard before voice and render: structure, spoken/display parity, triggers, doodles.

Errors make the result unusable (the renderer or timing would break); warnings are
quality notes. Both the rules director and LLM output must pass.
"""
from __future__ import annotations

import re
from pathlib import Path

from ..engine.storyboard import KINDS
from ..library import banned, resolve

SLOT_TYPES = {'cluster', 'quote', 'glossary', 'stat'}
PAGE_TYPES = {'ladder', 'bars', 'coins', 'grid100', 'lanes', 'range', 'zones', 'levels', 'table', 'dial', 'flow',
              'split', 'calendar'}
OTHER_TYPES = {'emphasis', 'stock'}
EN_PUNCT = re.compile(r'[,.;:?!](?=\s|$|["”’)])|—')
ZH_PUNCT = re.compile(r'[，。；：？！、—]')
MAX_SECTIONS = 8


def _triggers(value, path=''):
    """Yield (path, trigger dict) for every trigger in a visual, however deeply nested."""
    if isinstance(value, dict):
        for key, item in value.items():
            if key == 'trigger' and isinstance(item, dict):
                yield path, item
            else:
                yield from _triggers(item, f'{path}.{key}' if path else key)
    elif isinstance(value, list):
        for i, item in enumerate(value):
            yield from _triggers(item, f'{path}[{i}]')


def _doodles(value):
    if isinstance(value, dict):
        for key, item in value.items():
            if key == 'doodle' and isinstance(item, str):
                yield item
            else:
                yield from _doodles(item)
    elif isinstance(value, list):
        for item in value:
            yield from _doodles(item)


def validate(board: dict, project_dir: Path | None = None) -> dict:
    lang = board.get('lang')
    errors, warnings = [], []
    if lang not in ('en', 'zh'):
        return {'ok': False, 'errors': [f'lang must be en or zh, got {lang!r}'], 'warnings': []}
    chapters = board.get('chapters') or []
    ids = [c.get('id') for c in chapters]
    if len(ids) != len(set(ids)):
        errors.append('duplicate chapter ids')
    for c in chapters:
        if c.get('kind') not in KINDS:
            errors.append(f"chapter {c.get('id')}: kind {c.get('kind')!r} not in {sorted(KINDS)}")
    sections = [c for c in chapters if c.get('kind') == 'section']
    if len(sections) > MAX_SECTIONS:
        errors.append(f'{len(sections)} sections; the agenda holds at most {MAX_SECTIONS}')
    if sections and not any(c.get('kind') == 'agenda' for c in chapters):
        errors.append('sections need an agenda chapter')
    beats = board.get('beats') or []
    by_id = {b.get('id'): b for b in beats}
    if len(by_id) != len(beats):
        errors.append('duplicate beat ids')
    order = [b.get('chapter') for b in beats]
    runs = [c for i, c in enumerate(order) if i == 0 or order[i - 1] != c]
    if len(runs) != len(set(runs)):
        errors.append('beats of one chapter must be contiguous')
    if [c for c in ids if c in set(order)] != runs:
        errors.append('beat order does not follow chapter order')
    for c in chapters:
        if c.get('id') not in order:
            errors.append(f"chapter {c.get('id')} has no beats")
    visual_ids = set()
    punct = EN_PUNCT if lang == 'en' else ZH_PUNCT
    for b in beats:
        bid = b.get('id')
        spoken = (b.get('spoken') or {}).get(lang, '')
        display = (b.get('display') or {}).get(lang, '')
        if not spoken.strip() or not display.strip():
            errors.append(f'{bid}: empty spoken or display text')
            continue
        if re.search(r'\d', spoken):
            errors.append(f'{bid}: digits in spoken text ({spoken[:50]})')
        if punct.findall(spoken) != punct.findall(display):
            errors.append(f'{bid}: clause punctuation differs between spoken and display text')
        chapter = next((c for c in chapters if c.get('id') == b.get('chapter')), None)
        if chapter is None:
            errors.append(f'{bid}: unknown chapter {b.get("chapter")}')
            continue
        if b.get('kind') == 'take' and chapter.get('kind') == 'section':
            head = ((b.get('take') or {}).get('headline') or {}).get(lang, '')
            if not head:
                errors.append(f'{bid}: take beat without a headline')
            elif len(head.split() if lang == 'en' else head) > (16 if lang == 'en' else 30):
                warnings.append(f'{bid}: long takeaway headline')
        for v in b.get('visuals') or []:
            vid, vtype = v.get('id'), v.get('type')
            if not vid or vid in visual_ids:
                errors.append(f'{bid}: visual id missing or duplicated ({vid})')
            visual_ids.add(vid)
            if vtype not in SLOT_TYPES | PAGE_TYPES | OTHER_TYPES:
                errors.append(f'{bid}/{vid}: unknown visual type {vtype!r}')
            for where, trig in _triggers(v):
                ref = by_id.get(trig.get('beat'), b)
                phrase = trig.get(lang)
                if phrase and phrase not in (ref.get('spoken') or {}).get(lang, ''):
                    errors.append(f'{bid}/{vid} {where}: trigger {phrase!r} is not in the spoken text')
            for did in _doodles(v):
                if resolve(did, project_dir) is None:
                    errors.append(f'{bid}/{vid}: doodle {did!r} not found')
                elif did in banned()['doodles']:
                    warnings.append(f'{bid}/{vid}: {did!r} is a banned picture (no director picks it)')
            if vtype == 'cluster' and not 1 <= len(v.get('items') or []) <= 3:
                errors.append(f'{bid}/{vid}: a cluster needs 1-3 items')
    for b in beats:
        for v in b.get('visuals') or []:
            if v.get('type') == 'emphasis' and str(v.get('target', '')).split('.')[0] not in visual_ids:
                errors.append(f"{b['id']}/{v.get('id')}: emphasis target {v.get('target')!r} is unknown")
    return {'ok': not errors, 'errors': errors, 'warnings': warnings}

File: edu/extension/engine/validate.js
// Check a storyboard before voice and render: structure, spoken/display parity, triggers, doodles
// (port of doodlestudio/director/validate.py, English). Errors make it unusable; warnings are quality notes.
import { KINDS } from './storyboard.js';

const SLOT_TYPES = new Set(['cluster', 'quote', 'glossary', 'stat']);
export const PAGE_TYPES = new Set(['grid100', 'lanes', 'bars', 'flow', 'split']);      // the pages this renderer draws
const EN_PUNCT = /[,.;:?!](?=\s|$|["”’)])|—/g;
const MAX_SECTIONS = 8;

function* triggers(value, path = '') {
  if (Array.isArray(value)) {
    for (let i = 0; i < value.length; i++) yield* triggers(value[i], `${path}[${i}]`);
  } else if (value && typeof value === 'object') {
    for (const [key, item] of Object.entries(value)) {
      if (key === 'trigger' && item && typeof item === 'object' && !Array.isArray(item)) yield [path, item];
      else yield* triggers(item, path ? `${path}.${key}` : key);
    }
  }
}

function* doodles(value) {
  if (Array.isArray(value)) for (const item of value) yield* doodles(item);
  else if (value && typeof value === 'object') {
    for (const [key, item] of Object.entries(value)) {
      if (key === 'doodle' && typeof item === 'string') yield item;
      else yield* doodles(item);
    }
  }
}

/** board → { ok, errors, warnings }; `known(id)` says whether a doodle exists, `banned` is a Set of ids. */
export function validate(board, { known, banned = new Set() }) {
  const errors = [];
  const warnings = [];
  if (board.lang !== 'en') return { ok: false, errors: [`lang must be en, got ${board.lang}`], warnings };
  const chapters = board.chapters || [];
  const ids = chapters.map((c) => c.id);
  if (ids.length !== new Set(ids).size) errors.push('duplicate chapter ids');
  for (const c of chapters) if (!KINDS.has(c.kind)) errors.push(`chapter ${c.id}: kind ${c.kind} is unknown`);
  const sections = chapters.filter((c) => c.kind === 'section');
  if (sections.length > MAX_SECTIONS) errors.push(`${sections.length} sections; the agenda holds at most ${MAX_SECTIONS}`);
  if (sections.length && !chapters.some((c) => c.kind === 'agenda')) errors.push('sections need an agenda chapter');
  const beats = board.beats || [];
  const byId = new Map(beats.map((b) => [b.id, b]));
  if (byId.size !== beats.length) errors.push('duplicate beat ids');
  const order = beats.map((b) => b.chapter);
  const runs = order.filter((c, i) => i === 0 || order[i - 1] !== c);
  if (runs.length !== new Set(runs).size) errors.push('beats of one chapter must be contiguous');
  if (JSON.stringify(ids.filter((c) => order.includes(c))) !== JSON.stringify(runs)) errors.push('beat order does not follow chapter order');
  for (const c of chapters) if (!order.includes(c.id)) errors.push(`chapter ${c.id} has no beats`);
  const visualIds = new Set();
  for (const b of beats) {
    const spoken = b.spoken?.en || '';
    const display = b.display?.en || '';
    if (!spoken.trim() || !display.trim()) { errors.push(`${b.id}: empty spoken or display text`); continue; }
    if (/\d/.test(spoken)) errors.push(`${b.id}: digits in spoken text (${spoken.slice(0, 50)})`);
    if (JSON.stringify(spoken.match(EN_PUNCT) || []) !== JSON.stringify(display.match(EN_PUNCT) || [])) {
      errors.push(`${b.id}: clause punctuation differs between spoken and display text`);
    }
    const chapter = chapters.find((c) => c.id === b.chapter);
    if (!chapter) { errors.push(`${b.id}: unknown chapter ${b.chapter}`); continue; }
    if (b.kind === 'take' && chapter.kind === 'section') {
      const head = b.take?.headline?.en || '';
      if (!head) errors.push(`${b.id}: take beat without a headline`);
      else if (head.split(/\s+/).length > 16) warnings.push(`${b.id}: long takeaway headline`);
    }
    for (const v of b.visuals || []) {
      if (!v.id || visualIds.has(v.id)) errors.push(`${b.id}: visual id missing or duplicated (${v.id})`);
      visualIds.add(v.id);
      if (!SLOT_TYPES.has(v.type) && !PAGE_TYPES.has(v.type)) errors.push(`${b.id}/${v.id}: unknown visual type ${v.type}`);
      for (const [where, trig] of triggers(v)) {
        const ref = byId.get(trig.beat) || b;
        if (trig.en && !(ref.spoken?.en || '').includes(trig.en)) {
          errors.push(`${b.id}/${v.id} ${where}: trigger ${JSON.stringify(trig.en)} is not in the spoken text`);
        }
      }
      for (const id of doodles(v)) {
        if (!known(id)) errors.push(`${b.id}/${v.id}: doodle ${id} not found`);
        else if (banned.has(id)) warnings.push(`${b.id}/${v.id}: ${id} is a banned picture`);
      }
      if (v.type === 'cluster' && !((v.items || []).length >= 1 && (v.items || []).length <= 3)) {
        errors.push(`${b.id}/${v.id}: a cluster needs 1-3 items`);
      }
    }
  }
  return { ok: !errors.length, errors, warnings };
}
